### PR TITLE
fix: extremely rare race conditions in retry loop

### LINKED ISSUE #7788: Extremely unlikely race conditions in `internal::AsyncRetryLoop()`
Consider the current implementation of the asynchronous retry loop, in particular how the `pending_operation_` is set after the callback may have been invoked:

https://github.com/googleapis/google-cloud-cpp/blob/0c09efc51bc5c53b1930f25d7cefe07595bb5106/google/cloud/internal/async_retry_loop.h#L111-L115

If the callback is immediately completed (or the scheduler puts a long sleep before line 115) the `pending_operation_` may be set to a future that is no longer active, which means the cancel request has no effect:

https://github.com/googleapis/google-cloud-cpp/blob/0c09efc51bc5c53b1930f25d7cefe07595bb5106/google/cloud/internal/async_retry_loop.h#L192-L200

This is extremely rare, it would require multiple operations completing very fast, and separate threads going to sleep.  Nevertheless, we should fix this.

### PR DESCRIPTION
**Pull request by coryan**
Fixes #7788 

I could not write a unit test that would reliably introduce the race condition.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7789)

### CODE REVIEW

File: google/cloud/internal/async_retry_loop.h [hunk 1]
```diff
@@ -59,6 +59,109 @@ struct FutureValueType<future<T>> {
...
+ * One could solve this deadlock using a recursive mutex, but that still leaves
+ * a second problem: as the stack unwinds the `pending_operation_` member
+ * variable is set, on line (7), to the **first** operation, and we want it to
+ * remain set to the **last** operation.
+ *
+ * @par Solution
+ *
+ * We keep a counter reflecting the current iteration of the retry loop. This
+ * counter is incremented before starting a request and before starting a
+ * backoff timer.
```
Comment (dbolduc) [inline google/cloud/internal/async_retry_loop.h:118]: What do you think about changing all "iteration"s to "operation"? (e.g. `StartIteration()` -> `StartOperation()`, etc.)

"iteration" does makes sense when we think about the retry loop as `[request, backoff, request, backoff, request, ...]`, but I think "operation" is even clearer.

If you don't like that suggestion, maybe we could explicitly define the loop. Like: "The retry loop is an alternating sequence of requests and backoffs. We keep a counter reflecting the current iteration....". That way we won't think of one iteration as a request + a backoff.

**Review by dbolduc — COMMENTED** [review]
Logically, this LGTM. The documentation was helpful for understanding why we tag each operation with the iteration count.

File: google/cloud/internal/async_retry_loop.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/async_retry_loop.h]: Good idea, fixed.

File: google/cloud/internal/async_retry_loop.h [hunk 2]
```diff
@@ -59,6 +59,109 @@ struct FutureValueType<future<T>> {
...
+ * As to when the `cancelled_` flag changes to `true`:
+ *   - `StartIteration()` and `SetPending()` both lock the same mutex `mu_` as
+ *     `Cancel()`.
+ *   - It follows that if `Cancel()` is invoked, then the `true` value will be
+ *     visible any future calls to `StartIteration()` or `SetPending()`.
+ *   - If the next call is `StartIteration()` then no new operation is issued,
+ *     as both `StartAttempt()` and `StartBackoff()` return immediately in this
+ *     case.
+ *   - Note that if `cancelled_` is true, `StartIteration()` terminates the
+ *     retry loop by calling `SetDoneWithCancel()` if `cancelled_` is true.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_retry_loop.h]: Fixed.

File: google/cloud/internal/async_retry_loop.h [hunk 3]
```diff
@@ -218,9 +321,14 @@ class AsyncRetryLoopImpl
...
   promise<T> result_;
+
+  // Only the following variables require synchronization, as they coordinate
+  // the work between the retry loop (which would be lock-free) and the cancel
+  // requests (which need locks).
   std::mutex mu_;
-  State state_ = kIdle;
   bool cancelled_ = false;
+  bool done_ = false;
+  std::uint_fast32_t iteration_ = 0;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_retry_loop.h]: Thanks, fixed.

File: google/cloud/internal/async_retry_loop.h [hunk 4]
```diff
@@ -59,6 +59,109 @@ struct FutureValueType<future<T>> {
...
+ *     starts, as it is their callbacks who start the next step.
+ *   - `StartIteration()` is always sequenced-before calls to `SetPending()`.
+ *   - `SetPending()` never sets the `pending_operation_` to the future for
+ *     an operation that already completed.
+ *
+ * As to when the `cancelled_` flag changes to `true`:
+ *   - `StartIteration()` and `SetPending()` both lock the same mutex `mu_` as
+ *     `Cancel()`.
+ *   - It follows that if `Cancel()` is invoked, then the `true` value will be
+ *     visible any future calls to `StartIteration()` or `SetPending()`.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_retry_loop.h]: Done.

File: google/cloud/internal/async_retry_loop.h [hunk 5]
```diff
@@ -59,6 +59,109 @@ struct FutureValueType<future<T>> {
...
+ * - Once `cancelled_` is set to `true` it is never set to `false`
+ *
+ * While `cancelled_` is false the loop is (basically) single threaded:
+ *   - Each gRPC request or backoff timer is sequenced-after a call to
+ *     `StartIteration()`, see `StartAttempt()` and `StartBackoff()`.
+ *   - Each gRPC request or backoff timer must complete before the next one
+ *     starts, as it is their callbacks who start the next step.
+ *   - `StartIteration()` is always sequenced-before calls to `SetPending()`.
+ *   - `SetPending()` never sets the `pending_operation_` to the future for
+ *     an operation that already completed.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_retry_loop.h]: Done.

File: google/cloud/internal/async_retry_loop.h [hunk 6]
```diff
@@ -59,6 +59,109 @@ struct FutureValueType<future<T>> {
...
+ * The `pending_operation_` variable is updated **only** if the current
+ * iteration matches the iteration before the operation started. This means the
+ * `pending_operation_` variable always reflects the last pending operation, it
+ * can never be set to an older operation.
+ *
+ * @par Observations
+ *
+ * - The initial value of `cancelled_` is false.
+ * - `Cancel()` is the only operation that changes `cancelled_`, and it holds
+ *   the `mu_` mutex while doing son.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/async_retry_loop.h]: Done.

File: google/cloud/internal/async_retry_loop.h [hunk 7]
```diff
@@ -93,120 +197,120 @@ class AsyncRetryLoopImpl
...
-      case kDone:
-        break;
-      case kWaiting:
-        state_ = kIdle;
-        break;
-    }
-  }
-
-  void SetWaiting(future<void> op) {
+  void SetPending(std::uint_fast32_t iteration, future<void> op) {
```
Comment (dbolduc) [inline google/cloud/internal/async_retry_loop.h:280]: do you still want this parameter to be called `iteration`?

**Review by dbolduc — APPROVED** [review]

File: google/cloud/internal/async_retry_loop.h (hunk 7 above)
Comment (coryan) [reply google/cloud/internal/async_retry_loop.h]: Nope, fixed.

**Pull request merged**